### PR TITLE
Lab 05: implementation - Basanov Maxim

### PR DESCRIPTION
**Pull request by scruffyscarf**
# Lab 5 Pull Request

## Overview
- Implement jwt.go, password.go, and user.go
- Implement auth_service.dart and form_validator.dart

# PR with points
https://github.com/scruffyscarf/sum25-go-flutter-course/pull/9

### Go Backend Component
- [x] jwt.go
- [x] password.go
- [x] user.go

### Flutter Frontend Component
- [x] auth_service.dart
- [x] form_validator.dart

### Frontend Tests
```bash
cd labs/lab05/frontend
flutter pub get
flutter packages pub
flutter test
```

### Backend Tests
```bash
cd labs/lab05/backend
go mod tidy
go test ./...
```

## Checklist

- [x] All tests pass (go test ./... and flutter test)
- [x] Code passes linting (go vet ./... and dart analyze)
- [x] Both Go and Flutter components work
- [x] Follows course coding standards

### CODE REVIEW

File: labs/lab05/backend/security/password.go [hunk 1]
```diff
@@ -2,54 +2,54 @@ package security
...
-// TODO: Implement ValidatePassword function
-// ValidatePassword checks if password meets basic requirements
-// Requirements:
-// - At least 6 characters
-// - Contains at least one letter and one number
 func ValidatePassword(password string) error {
-	// TODO: Implement password validation
-	// Check length and basic complexity requirements
-	return errors.New("not implemented")
+	if len(password) < 6 {
```
Comment (wrzdx) [inline labs/lab05/backend/security/password.go:36]: Consider using a single compiled regexp for performance, especially if this function is called frequently. Also, think about whether to enforce additional password complexity (e.g., special characters) or make the minimum length configurable.

File: labs/lab05/backend/userdomain/user.go [hunk 2]
```diff
@@ -2,70 +2,95 @@ package userdomain
...
 func ValidateEmail(email string) error {
-	// TODO: Implement email validation
-	// Use regex pattern to validate email format
-	// Email should not be empty and should match standard email pattern
-	return errors.New("not implemented")
+	trimmed := strings.TrimSpace(email)
+	if trimmed == "" {
+		return errors.New("email cannot be empty")
+	}
+	pattern := `^[a-zA-Z0-9._%+-]+@[a-zA-Z0-9.-]+\.[a-zA-Z]{2,}$`
```
Comment (wrzdx) [inline labs/lab05/backend/userdomain/user.go:55]: Basic regex covers most emails, but you might want to consider more robust validation or using a well-tested library to handle edge cases in email formatting.

**Review by wrzdx — COMMENTED** [review]
This PR implements core backend and frontend validation and authentication logic with clear and readable code. Good use of input validation and error handling is evident throughout. However, there are areas where robustness could be improved, such as combining or reusing regular expressions for efficiency. For future improvements, consider whether validation rules should be centralized and configurable. Overall, this is a solid foundation—tightening validation will further strengthen the codebase.

**Review by NikitaKh06 — COMMENTED** [review]
The return nil, errors.New("not implemented") statements should be removed from the beginning of the methods; they make the code unreachable.

In ValidateToken, the alg field from the token header is passed to NewInvalidSigningMethodError — better to use token.Method.Alg() for accuracy and type safety.

Consider logging the actual err on parsing for better diagnostics.

NewPasswordService() should not return nil initially.

The validation rule (at least one letter and one number) is clearly enforced, but consider:

making the rules configurable;

providing better errors (e.g., "missing letter" vs. "missing digit").

All validation helper functions still contain return errors.New("not implemented") above the real logic — these should be removed.

Consider unifying password validation logic between security/password.go and userdomain/user.go to avoid duplication.


As with the backend, throw UnimplementedError(...) remains at the top of several completed methods — these must be removed to enable actual execution.

Error messages can be standardized (e.g., use capitalized phrases consistently or provide field-specific context).
